File: java/com/google/copybara/git/ChangeReader.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git;

import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.base.Joiner;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.copybara.Change;
import com.google.copybara.ChangeMessage;
import com.google.copybara.authoring.Author;
import com.google.copybara.authoring.Authoring;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.git.GitRepository.GitLogEntry;
import com.google.copybara.git.GitRepository.LogCmd;
import com.google.copybara.util.Glob;
import com.google.copybara.util.console.Console;
import java.util.List;
import java.util.stream.Collectors;
import javax.annotation.Nullable;

/**
 * Utility class to introspect the log of a Git repository.
 */
class ChangeReader {

  @Nullable private final Authoring authoring;
  private final GitRepository repository;
  private final int limit;
  private final ImmutableList<String> roots;
  private final boolean includeBranchCommitLogs;
  private final String url;
  private final boolean firstParent;
  private final boolean partialFetch;
  private final int skip;
  @Nullable private final String grepString;

  private ChangeReader(@Nullable Authoring authoring, GitRepository repository, int limit,
      Iterable<String> roots, boolean includeBranchCommitLogs, @Nullable String url,
      boolean firstParent, boolean partialFetch, int skip, @Nullable String grepString) {
    this.authoring = authoring;
    this.repository = checkNotNull(repository, "repository");
    this.limit = limit;
    this.roots = ImmutableList.copyOf(roots);
    this.includeBranchCommitLogs = includeBranchCommitLogs;
    this.url = url;
    this.firstParent = firstParent;
    this.partialFetch = partialFetch;
    this.skip = skip;
    this.grepString = grepString;
  }

  ImmutableList<Change<GitRevision>> run(String refExpression)
      throws RepoException, ValidationException {
    LogCmd logCmd = repository
        .log(refExpression)
        .firstParent(firstParent);
    if (limit != -1) {
      logCmd = logCmd.withLimit(limit);
    }
    if (skip > 0) {
      logCmd = logCmd.withSkip(skip);
    }
    if (grepString != null) {
      logCmd = logCmd.grep(grepString);
    }
    if (partialFetch && roots.contains("")) {
      throw new ValidationException("Config error: partial_fetch feature is not compatible "
          + "with fetching the whole repo.");
    }
    if (partialFetch) {
      logCmd = logCmd.withPaths(roots);
    }
    // Log command does not filter by roots here because of how git log works. Some commits (e.g.
    // fake merges) might not include the files in the log, and filtering here would return
    // incorrect results. We do filter later on the changes to match the actual glob.
    return parseChanges(logCmd.includeFiles(true).includeMergeDiff(true).run());
  }

  static final String BRANCH_COMMIT_LOG_HEADING = "-- Branch commit log --";

  private CharSequence branchCommitLog(GitRevision ref, List<GitRevision> parents)
      throws RepoException {
    if (parents.size() <= 1) {
      // Not a merge commit, so don't bother showing full log of branch commits. This would only
      // contain the raw commit of 'ref', which will be redundant.
      return "";
    }
    if (!includeBranchCommitLogs) {
      return "";
    }

    ImmutableList<GitLogEntry> entries =
        repository
            .log(parents.get(0).getSha1() + ".." + ref.getSha1())
            // This might give incorrect results but several migrations rely on this behavior.
            // and first_parent = False doesn't work for ITERATIVE
            .withPaths(Glob.isEmptyRoot(roots) ? ImmutableList.of() : roots)
            .firstParent(false)
            .run();

    if (entries.isEmpty()) {
      return "";
    }
    // Remove the merge commit. Since we already have that in the body.
    entries = entries.subList(1, entries.size());

    return "\n" + BRANCH_COMMIT_LOG_HEADING + "\n" +
        Joiner.on("\n").join(entries.stream()
            .map(e -> ""
                + "commit " + e.getCommit().getSha1() + "\n"
                + "Author:  " + filterAuthor(e.getAuthor()) + "\n"
                + "Date:    " + e.getAuthorDate() + "\n"
                + "\n"
                + "    " + e.getBody().replace("\n", "    \n"))
            .collect(Collectors.toList()));
  }

  private ImmutableList<Change<GitRevision>> parseChanges(ImmutableList<GitLogEntry> logEntries)
      throws RepoException {

    ImmutableList.Builder<Change<GitRevision>> result = ImmutableList.builder();
    GitRevision last = null;
    for (GitLogEntry e : logEntries) {
      // Keep the first commit if repeated (merge commits).
      if (last != null && last.equals(e.getCommit())) {
        continue;
      }
      last = e.getCommit();
      result.add(new Change<>(
          e.getCommit().withUrl(url),
          filterAuthor(e.getAuthor())
          , e.getBody() + branchCommitLog(e.getCommit(), e.getParents()),
          e.getAuthorDate(),
          ChangeMessage.parseAllAsLabels(e.getBody()).labelsAsMultimap(),
          e.getFiles(), e.getParents().size() > 1, e.getParents()));
    }
    return result.build().reverse();
  }

  private Author filterAuthor(Author author) {
    return authoring == null || authoring.useAuthor(author.getEmail())
        ? author
        : authoring.getDefaultAuthor();
  }

  /**
   * Builder for ChangeReader.
   */
  static class Builder {

    private Authoring authoring = null;
    private final GitRepository repository;
    private int limit = -1;
    private ImmutableList<String> roots = ImmutableList.of("");
    private boolean includeBranchCommitLogs = false;
    private String url;
    private boolean firstParent;
    private boolean partialFetch;
    private int skip;
    private String grepString;

    // TODO(matvore): Consider adding destinationFiles.
    // For ALL_FILES and where roots is [""], This will skip merges that don't affect the tree
    // For other cases, this will skip merges and commits that don't affect a subtree
    static Builder forDestination(GitRepository repository, Console console) {
      return new Builder(repository, console);
    }

    static Builder forOrigin(Authoring authoring, GitRepository repository, Console console) {
      return new Builder(repository, console)
          .setAuthoring(authoring);
    }

    private Builder(GitRepository repository, Console console) {
      this.repository = checkNotNull(repository, "repository");
      checkNotNull(console, "console");
    }

    Builder setLimit(int limit) {
      Preconditions.checkArgument(limit > 0);
      this.limit = limit;
      return this;
    }

    Builder setSkip(int skip) {
      Preconditions.checkArgument(skip >= 0);
      this.skip = skip;
      return this;
    }

    private Builder setAuthoring(Authoring authoring) {
      this.authoring = checkNotNull(authoring, "authoring");
      return this;
    }

    Builder setPartialFetch(boolean partialFetch) {
      this.partialFetch = partialFetch;
      return this;
    }

    Builder setFirstParent(boolean firstParent) {
      this.firstParent = firstParent;
      return this;
    }

    Builder setIncludeBranchCommitLogs(boolean includeBranchCommitLogs) {
      this.includeBranchCommitLogs = includeBranchCommitLogs;
      return this;
    }

    Builder setUrl(String url) {
      this.url = url;
      return this;
    }

    /**
     * Only return commits that match the given paths in the Git log command.
     */
    Builder setRoots(Iterable<String> roots) {
      this.roots = ImmutableList.copyOf(roots);
      return this;
    }

    /**
     * Grep for the given pattern in the Git log command.
     */
    Builder grep(String grepString) {
      this.grepString = grepString;
      return this;
    }

    ChangeReader build() {
      return new ChangeReader(
          authoring, repository, limit, roots, includeBranchCommitLogs, url,
          firstParent, partialFetch, skip, grepString);
    }
  }

}

File: src/Copybara.Core/Git/ChangeReader.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Text;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Revision;
using Copybara.Util;
using Console = Copybara.Util.Console.Console;
using AuthoringT = Copybara.Authoring.Authoring;
using Author = Copybara.Authoring.Author;

namespace Copybara.Git;

/// <summary>
/// Utility class to introspect the log of a Git repository. Port of
/// <c>com.google.copybara.git.ChangeReader</c>.
/// </summary>
internal sealed class ChangeReader
{
    internal const string BranchCommitLogHeading = "-- Branch commit log --";

    private readonly AuthoringT? _authoring;
    private readonly GitRepository _repository;
    private readonly int _limit;
    private readonly ImmutableArray<string> _roots;
    private readonly bool _includeBranchCommitLogs;
    private readonly string? _url;
    private readonly bool _firstParent;
    private readonly bool _partialFetch;
    private readonly bool _topoOrder;
    private readonly int _skip;
    private readonly int _batchSize;
    private readonly string? _grepString;

    private ChangeReader(
        AuthoringT? authoring,
        GitRepository repository,
        int limit,
        IEnumerable<string> roots,
        bool includeBranchCommitLogs,
        string? url,
        bool firstParent,
        bool partialFetch,
        bool topoOrder,
        int skip,
        int batchSize,
        string? grepString)
    {
        _authoring = authoring;
        _repository = Preconditions.CheckNotNull(repository, "repository");
        _limit = limit;
        _roots = roots.ToImmutableArray();
        _includeBranchCommitLogs = includeBranchCommitLogs;
        _url = url;
        _firstParent = firstParent;
        _partialFetch = partialFetch;
        _topoOrder = topoOrder;
        _skip = skip;
        _batchSize = batchSize;
        _grepString = grepString;
    }

    internal IReadOnlyList<Change<GitRevision>> Run(GitRevision rev) =>
        Run(
            null,
            rev,
            historyIsNonLinear: false,
            ImmutableDictionary<string, ImmutableListMultimap<string, string>>.Empty);

    /// <summary>
    /// Computes a list of changes from the refExpression and propagates labels contained in
    /// <paramref name="labels"/> onto the resulting change list.
    /// </summary>
    internal IReadOnlyList<Change<GitRevision>> Run(
        GitRevision? fromRev,
        GitRevision toRev,
        bool historyIsNonLinear,
        IReadOnlyDictionary<string, ImmutableListMultimap<string, string>> labels)
    {
        string refExpression =
            fromRev == null || historyIsNonLinear
                ? toRev.GetHash()
                : fromRev.GetHash() + ".." + toRev.GetHash();
        GitRepository.LogCmd logCmd =
            _repository.Log(refExpression).FirstParent(_firstParent).TopoOrder(_topoOrder);
        if (_limit != -1)
        {
            logCmd = logCmd.WithLimit(_limit);
        }
        if (_skip > 0)
        {
            logCmd = logCmd.WithSkip(_skip);
        }
        if (_batchSize > 0)
        {
            logCmd = logCmd.WithBatchSize(_batchSize);
        }
        if (_grepString != null)
        {
            logCmd = logCmd.Grep(_grepString);
        }
        if (_partialFetch && _roots.Contains(""))
        {
            throw new ValidationException(
                "Config error: partial_fetch feature is not compatible with fetching the whole"
                    + " repo.");
        }
        if (_partialFetch)
        {
            logCmd = logCmd.WithPaths(_roots);
        }

        // Log command does not filter by roots here because of how git log works. Some commits (e.g.
        // fake merges) might not include the files in the log, and filtering here would return
        // incorrect results. We do filter later on the changes to match the actual glob.
        return ParseChanges(
            logCmd.IncludeFiles(true).IncludeMergeDiff(true).Run(), labels, toRev);
    }

    private string BranchCommitLog(GitRevision @ref, IReadOnlyList<GitRevision> parents)
    {
        if (parents.Count <= 1)
        {
            // Not a merge commit, so don't bother showing full log of branch commits.
            return "";
        }
        if (!_includeBranchCommitLogs)
        {
            return "";
        }

        IReadOnlyList<GitRepository.GitLogEntry> entries =
            _repository
                .Log(parents[0].GetHash() + ".." + @ref.GetHash())
                .WithPaths(Glob.IsEmptyRoot(_roots) ? ImmutableArray<string>.Empty : _roots)
                .FirstParent(false)
                .Run();

        if (entries.Count == 0)
        {
            return "";
        }
        // Remove the merge commit. Since we already have that in the body.
        entries = entries.Skip(1).ToList();

        var sb = new StringBuilder("\n").Append(BranchCommitLogHeading).Append('\n');
        bool first = true;
        foreach (var e in entries)
        {
            if (!first)
            {
                sb.Append('\n');
            }
            sb.Append("commit ")
                .Append(e.Commit.GetHash())
                .Append('\n')
                .Append("Author:  ")
                .Append(FilterAuthor(e.Author))
                .Append('\n')
                .Append("Date:    ")
                .Append(e.AuthorDate)
                .Append('\n')
                .Append('\n')
                .Append("    ")
                .Append((e.Body ?? "").Replace("\n", "    \n"));
            first = false;
        }
        return sb.ToString();
    }

    private IReadOnlyList<Change<GitRevision>> ParseChanges(
        IReadOnlyList<GitRepository.GitLogEntry> logEntries,
        IReadOnlyDictionary<string, ImmutableListMultimap<string, string>> labels,
        GitRevision toRev)
    {
        var result = new List<Change<GitRevision>>();
        GitRevision? last = null;
        foreach (var e in logEntries)
        {
            // Keep the first commit if repeated (merge commits).
            if (last != null && last.Equals(e.Commit))
            {
                continue;
            }
            last = e.Commit;
            ImmutableListMultimap<string, string> labelsToCopy =
                labels.TryGetValue(e.Commit.GetHash(), out var found)
                    ? found
                    : ImmutableListMultimap<string, string>.Empty;
            // Carry over the context reference to the corresponding change in the list.
            if (last.GetHash() == toRev.GetHash() && toRev.ContextReference() != null)
            {
                last = last.WithContextReference(toRev.ContextReference()!);
            }
            result.Add(
                new Change<GitRevision>(
                    (_url != null ? last.WithUrl(_url) : last).WithLabels(labelsToCopy),
                    FilterAuthor(e.Author),
                    (e.Body ?? "") + BranchCommitLog(last, e.Parents),
                    e.AuthorDate,
                    ChangeMessage.ParseAllAsLabels(e.Body ?? "").LabelsAsMultimap(),
                    e.Files,
                    e.Parents.Count > 1,
                    e.Parents.ToImmutableArray()));
        }
        result.Reverse();
        return result;
    }

    private Author FilterAuthor(Author author) =>
        _authoring == null || _authoring.UseAuthor(author.Email)
            ? author
            : _authoring.GetDefaultAuthor();

    /// <summary>Builder for <see cref="ChangeReader"/>.</summary>
    internal sealed class Builder
    {
        private AuthoringT? _authoring;
        private readonly GitRepository _repository;
        private int _limit = -1;
        private ImmutableArray<string> _roots = ImmutableArray.Create("");
        private bool _includeBranchCommitLogs;
        private string? _url;
        private bool _firstParent;
        private bool _topoOrder;
        private bool _partialFetch;
        private int _skip;
        private int _batchSize;
        private string? _grepString;

        internal static Builder ForDestination(GitRepository repository, Console console) =>
            new(repository, console);

        internal static Builder ForOrigin(
            AuthoringT authoring, GitRepository repository, Console console) =>
            new Builder(repository, console).SetAuthoring(authoring);

        private Builder(GitRepository repository, Console console)
        {
            _repository = Preconditions.CheckNotNull(repository, "repository");
            Preconditions.CheckNotNull(console, "console");
        }

        internal Builder SetLimit(int limit)
        {
            Preconditions.CheckArgument(limit > 0);
            _limit = limit;
            return this;
        }

        internal Builder SetSkip(int skip)
        {
            Preconditions.CheckArgument(skip >= 0);
            _skip = skip;
            return this;
        }

        internal Builder SetBatchSize(int batchSize)
        {
            Preconditions.CheckArgument(batchSize >= 0);
            _batchSize = batchSize;
            return this;
        }

        private Builder SetAuthoring(AuthoringT authoring)
        {
            _authoring = Preconditions.CheckNotNull(authoring, "authoring");
            return this;
        }

        internal Builder SetPartialFetch(bool partialFetch)
        {
            _partialFetch = partialFetch;
            return this;
        }

        internal Builder SetFirstParent(bool firstParent)
        {
            _firstParent = firstParent;
            return this;
        }

        internal Builder SetTopoOrder(bool topoOrder)
        {
            _topoOrder = topoOrder;
            return this;
        }

        internal Builder SetIncludeBranchCommitLogs(bool includeBranchCommitLogs)
        {
            _includeBranchCommitLogs = includeBranchCommitLogs;
            return this;
        }

        internal Builder SetUrl(string? url)
        {
            _url = url;
            return this;
        }

        /// <summary>Only return commits that match the given paths in the Git log command.</summary>
        internal Builder SetRoots(IEnumerable<string> roots)
        {
            _roots = roots.ToImmutableArray();
            return this;
        }

        /// <summary>Grep for the given pattern in the Git log command.</summary>
        internal Builder Grep(string grepString)
        {
            _grepString = grepString;
            return this;
        }

        internal ChangeReader Build() =>
            new(
                _authoring,
                _repository,
                _limit,
                _roots,
                _includeBranchCommitLogs,
                _url,
                _firstParent,
                _partialFetch,
                _topoOrder,
                _skip,
                _batchSize,
                _grepString);
    }
}
